deltametrics/io.py

```diff
@@ -0,0 +1,217 @@
+
+        This function should initialize the file if it does not exist, or
+        connect to the file if it already exists---but *do not* read the file.
+        """
+        return
+
+    @abc.abstractmethod
+    def read(self):
+        """Should read data into memory.
+        """
+        return
+
+    @abc.abstractmethod
+    def write(self):
+        """Should write the data to file.
+
+        Take a :obj:`~deltametrics.cube.Cube` and write it to file.
+        """
+        return
+
+    @abc.abstractmethod
+    def __getitem__(self):
+        """Should slice the data from file.
+        """
+        return
+
+    @property
+    @abc.abstractmethod
+    def keys(self):
+        """Should link to all key _names_ stored in file.
+        """
+        return
+
+
+class NetCDFIO(BaseIO):
+    """Utility for consistent IO with netCDF files.
+
+    This module wraps calls to the netCDF4 python module in a consistent API,
+    so the user can work seamlessly with either netCDF4 files or HDF5 files.
+    The public methods of this class are consistent with
+    :obj:`~deltametrics.utils.HDFIO`.
+    """
+
+    def __init__(self, data_path, write=False):
+        """Initialize the NetCDFIO handler.
+
+        Initialize a connection to a NetCDF file.
+
+        Parameters
+        ----------
+        data_path : `str`
+            Path to file to read or write to.
+
+        write : `bool`, optional
+            Whether to allow writing to an existing file. Set to False by
+            default, if a file already exists at ``data_path``, writing is
+            disabled, unless ``write`` is set to True.
+        """
+
+        super().__init__(data_path=data_path, write=write)
+
+        self.type = 'netcdf'
+        self._in_memory_data = {}
+
+    def connect(self):
+        """Connect to the data file.
+
+        Initialize the file if it does not exist, or simply ``return`` if the
+        file already exists.
+
+        .. note::
+            This function is automatically called during initialization of any
+            IO object, so it is not necessary to call it directly.
+
+        """
+        if not os.path.isfile(self.data_path):
+            self.dataset = netCDF4.Dataset(
+                self.data_path, "w", format="NETCDF4")
+        else:
+            if self.write:
+                self.dataset = netCDF4.Dataset(self.data_path, "r+")
+            else:
+                self.dataset = netCDF4.Dataset(self.data_path, "r")
+
+    def read(self, var):
+        """Read variable from file and into memory.
+
+        Converts `variables` in netCDF file to `ndarray` for coersion into a
+        :obj:`~deltametrics.cube.Cube` instance.
+
+        Parameters
+        ----------
+        var : `str`
+            Which variable to load from the file.
+        """
+        try:
+            _arr = self.dataset.variables[var]
+        except ValueError as e:
+            raise e
+        self._in_memory_data[var] = np.array(_arr, copy=True)
```

Comment:
Is there a way to warn the person if they're about to load a really large dataset into memory? Say, it the array is going to be >4 GB of ram, just raise a warning to make sure they know?

Reply:
Sure, I can throw a warning in there. Would you like it to be a "reject" warning, where we require a `force=True` to override the warning and load anyway?

The reading is disabled by default though, so if someone tries to read a 4GB file...well, swim at your own risk...

Reply:
Lol. Yeah I agree that people shouldn't do it unless they know what they're doing, I foresee this being a thing that someone might do while exploring the API, and just try to read all variables from a 50 GB dataset into memory, without realizing it. I think a `force=True` warning would be a great way to put this in there, potentially with a short call to `psutil` at the beginning to query the system and figure out what the total memory capacity is, and only throw a warning if the file is, say >80% of the total available.